### PR TITLE
Add validateWith function

### PR DESCRIPTION
**Pull request by rpearce**
## The Problem
Can't use `validate` with a joi object that has been extended (see https://github.com/hapijs/joi/blob/v14.0.1/API.md#extendextension)

## This Solution
~Make a breaking change whereby a joi object is required as the first argument, enabling any joi object – extended or not – to be utilized.~

Add a `validateWith` function that takes in a `joi` object as its first argument.

## Side Benefit
As you can see in the tests, we can now compose extensions and validate with them.

### CODE REVIEW

**Comment by rpearce** [comment]
Alrighty, after getting some feedback, I've updated this to not replace `validate` but be another function,`validateWith`, in order to avoid a breaking change (major version bump).

File: src/validateWith.js [hunk 1]
```diff
@@ -0,0 +1,16 @@
+const curryN = require('ramda/src/curryN')
+
+const promisify = require('./promisify')
+const defaults  = { abortEarly: false }
+
+const validateWith = (joi, schema, x, opts=defaults) => {
```
[the remark this answers is not in the document]
Comment (evilsoft) [reply src/validateWith.js]: And I see that it is, looking at `validate`. Please ignore me!!

File: src/validateWith.js [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+  try {
+    const _validateWith = promisify(joi.validate, joi)
+    return _validateWith(x, schema, opts)
+  } catch (_) {
```
[the remark this answers is not in the document]
Comment (rpearce) [reply src/validateWith.js]: I did this to not blow up in the following cases (which maybe we shouldn't?):
  • `null`/`undefined`/??? is passed as the first argument (could run checks? the next one is trickier)
  • `{ validate: () => 'wrong object that happens to have validate method' }` or some such is passed as the first argument

File: test/validateWith.js [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+    )
+  })
+
+  describe('when joi is missing', () => {
+    beforeEach(() =>
+      validateWith(null, schema, bad).then(res)
+    )
+
+    it('pass-thrus values without validation', () =>
+      expect(res()).to.eql(bad)
```
[the remark this answers is not in the document]
Comment (rpearce) [reply test/validateWith.js]: Goes with the `try/catch` in the other comment: if `validateWith` completely borks, do we want to resolve or reject? I tried following https://github.com/articulate/funky/blob/master/src/validate.js#L17 which does the same thing when `joi` is borked (though for different reasons)

File: src/validateWith.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (mgreystone) [reply src/validateWith.js]: IMHO, we should just allow :boom: in this case. (It's the ramda way...) If you don't pass valid arguments, an implicit "pass" sounds dangerous.

[the remark this answers is not in the document]
Comment (rpearce) [reply src/validateWith.js]: Works for me! Let it 🔥

File: test/validateWith.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (mgreystone) [reply test/validateWith.js]: I'd argue against it here too. If i forget to `npm install joi` on my project, i'd have for all my validations to pass unexpectedly. That is a out-of-scope breaking change, though...

**Comment by rpearce** [comment]
Sweet! What's releasing look like for this project? Is it...
1. merge
1. ...wait for other things if more coming...
1. determine major/minor/patch (_minor_ for this b/c new, backward-compatible functionality)
1. _(don't see a CHANGELOG)_
1. create a GH release
1. `npm publish`

**Comment by flintinatux** [comment]
@rpearce:
1. merge
2. `git checkout master && git pull origin master`
3. `yarn version` -- enter the next minor version number
4. `npm publish --access=public`

**Comment by rpearce** [comment]
Ty

**Pull request merged**